third_party/nginx_latency_status_module/src/ngx_http_latency_stub_status_module.c

```diff
@@ -0,0 +1,305 @@
+// modification, are permitted provided that the following conditions
+// are met:
+// 1. Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+// 2. Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+// OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+// SUCH DAMAGE.
+////////////////////////////////////////////////////////////////////////////////
+
+#include "ngx_http_latency_recorder.h"
+#include "ngx_http_latency_status_handler.h"
+#include "ngx_http_latency_storage.h"
+#include "ngx_http_latency_stub_status_module.h"
+
+// A variable used to get the max_exponent from the main config in the
+// shared memory initialization
+static ngx_int_t shm_max_exponent;
+
+
+// Initialize a single set of latency metrics.
+latency_stat *create_latency_record(ngx_slab_pool_t *shpool) {
+  latency_stat *record;
+
+  record = ngx_slab_alloc(shpool, sizeof(latency_stat));
+  if (record == NULL) {
+    return record;
+  }
+
+  record->request_count = 0;
+  record->sum = 0;
+  record->sum_squares = 0;
+
+  record->distribution = ngx_slab_calloc(shpool, sizeof(ngx_atomic_t) * (shm_max_exponent + 2));
+  if (record->distribution == NULL) {
+    ngx_slab_free(shpool, record);
+    return NULL;
+  }
+
+  return record;
+}
+
+
+// Initialize the shared memory used to store the latency statistics.
+ngx_int_t ngx_http_latency_init_shm_zone(ngx_shm_zone_t *shm_zone, void *data){
+  ngx_slab_pool_t *shpool;
+  ngx_http_latency_shm_t *record_set;
+
+  if (data) {
+    shm_zone->data = data;
+    return NGX_OK;
+  }
+
+  shpool = (ngx_slab_pool_t *)shm_zone->shm.addr;
+  record_set = ngx_slab_alloc(shpool, sizeof(ngx_http_latency_shm_t));
+
+  record_set->request_latency = create_latency_record(shpool);
+  record_set->upstream_latency = create_latency_record(shpool);
+  record_set->websocket_latency = create_latency_record(shpool);
+
+  if (record_set->request_latency == NULL ||
+      record_set->upstream_latency == NULL ||
+      record_set->websocket_latency == NULL) {
+    return NGX_ERROR;
+  }
```

Comment:
Thanks for addressing my `free` concern above. Similar to it, could we split the null checks after each `create_latency_record` so that we can `free` the unsuccessfully allocated record? So:

record_set->request_latency = ...
if (record_set->request_latency = NULL) {
    return
}
record_set->upstream_latency = ...
if (record_set->upstream_latency = NULL) {
    free request_latency
    return
}
...

Reply:
done